web/server.go

```diff
@@ -139,38 +145,81 @@ func ListenAndServe() (*Servers, error) {
 		return nil, err
 	}
 
-	return &Servers{
-		major: major,
-		admin: admin,
-	}, nil
+	servers := NewServers()
+	servers.Start(major, "major", config.ServerAddrs())
+	servers.Start(admin, "admin", config.AdminServerAddrs())
+
+	return servers, nil
 }
 
-// Servers contains the started HTTP servers and implement the Shutdowner
-// interface.
+// Servers allow to start several [echo.Echo] servers and stop them together.
+//
+//	It also take care of several other task:
+//	- It sanitize the hosts format
+//	- It exposes the handlers on several addresses if needed
+//	- It forces the IPv4/IPv6 dual stack mode for `localhost` by
+//	  remplacing this entry by `["127.0.0.1", "::1]`
 type Servers struct {
-	major *echo.Echo
-	admin *echo.Echo
-	errs  chan error
+	servers []utils.Shutdowner
+	errs    chan error
 }
 
-// Start starts the servers.
-func (e *Servers) Start() {
-	e.errs = make(chan error)
+func NewServers() *Servers {
+	return &Servers{
+		servers: []utils.Shutdowner{},
+		errs:    make(chan error),
+	}
+}
+
+// Start the server 'e' to the given addrs.
+//
+// The 'addrs' arguments must be in the format `"host:port"`. If the host
+// is not a valid IPv4/IPv6/hostname or if the port not present an error is
+// returned.
+func (s *Servers) Start(e *echo.Echo, name string, addrs []string) error {
+	parsed := map[string]string{}
+
+	if len(addrs) == 0 {
+		return fmt.Errorf("args: %w", ErrMissingArgument)
+	}
 
-	go e.start(e.major, "major", &http.Server{
-		Addr:              config.ServerAddr(),
-		ReadHeaderTimeout: ReadHeaderTimeout,
-	})
+	if len(name) == 0 {
+		return fmt.Errorf("name: %w", ErrMissingArgument)
+	}
 
-	go e.start(e.admin, "admin", &http.Server{
-		Addr:              config.AdminServerAddr(),
-		ReadHeaderTimeout: ReadHeaderTimeout,
-	})
-}
+	for _, addr := range addrs {
+		host, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			return err
+		}
+
+		if host == "localhost" {
+			parsed["127.0.0.1"] = port
+			parsed["::1"] = port
+			continue
+		}
+
+		parsed[host] = port
+	}
 
-func (e *Servers) start(s *echo.Echo, name string, server *http.Server) {
-	fmt.Printf("  http server %s started on %q\n", name, server.Addr)
-	e.errs <- s.StartServer(server)
+	for host, port := range parsed {
+		addr := net.JoinHostPort(host, port)
+
+		fmt.Printf("http server %s started on %q\n", name, addr)
+		go func(addr string) {
+			s.errs <- e.StartServer(&http.Server{
+				Addr:              addr,
+				ReadHeaderTimeout: ReadHeaderTimeout,
+			})
+		}(addr)
+	}
+
+	s.servers = append(s.servers, e)
+
+	// Need some time to start the goroutines.
+	time.Sleep(50 * time.Millisecond)
```

Comment:
This can avoid panic when calling `servers.Shutdown` before the last server is started. This is a case I have encountered inside the tests for example. 

That's really not a "must have" and I can move the sleeps back inside the tests. I'm not a fan of the sleep solution either.

Reply:
Correction: This is not necessary anymore, the panic come from the situation where we consume the first event in `servers.errs` by calling `servers.Wait()` then we call directly `Shutdown` and we never consume the remaining errors events when the other servers are closing. As I tried to close the channel and a server tried to push an event at the same time, it panics. It will solve this in another PR